### PR TITLE
feat(core): Add certificates to Operator.Web

### PR DESCRIPTION
**Pull request by ian-buse**
Hey there @buehler!

This PR adds a new CertificateGenerator to Operator.Web, and it reworks a handful of the classes in Operator.Web to support using certificates in code. I more or less re-implemented the BouncyCastle solution from the CLI in Operator.Web using the built-in .NET classes, and I went to some length to ensure the generated certificates are virtually identical to those produced by BouncyCastle.

As for the reasoning, I admittedly had some trouble getting LocalTunnel working reliably in my environment. My company prefers just using our internal network versus a service, but I wanted to have a easy way to use self-signed certs with Kestrel and using your nifty webhook generator methods that are in the `DevelopmentTunnelService`. It also requires some knowledge of certificates and keys to get the BouncyCastle library working with certain .NET methods/classes, so this was a bit of a usability barrier for those who are unfamiliar.

Here are the changes I've made:
- Pulled the webhook generator methods into a base class. The base class is implemented by `TunnelWebhookService` and `CertificateWebhookService`. The new `CertificateWebhookService` is added to the operator with a new method in `OperatorBuilderExtensions`, called `UseCertificateProvider()`.
  - Moved or renamed a couple of classes to support this (`TunnelConfig` -> `WebhookConfig`, `WebhookLoader` moved).
- Created a new `CertificateGenerator` class in Operator.Web.
  - Added some supporting extension methods.
  - The `CertificateGenerator` implements an interface called `ICertificateProvider`, which is used by `UseCertificateProvider()` so someone can implement their own solution (like reading from a file) if they want.
- I replaced BouncyCastle in the CLI with the Operator.Web generator in 3bb58fb.
- I added a few tests in ce7a7b5.

There are no breaking changes with these commits. 

Regards,

Ian

### CODE REVIEW

**Comment by buehler** [comment]
Hey @ian-buse 

Thank you for your contribution! :-)

Yes, certificate stuff is quite weird, and you need a lot of time to get your head around it. The way these certificates are implemented now is to ensure that webservers do not reject them because they all need certain custom extensions.

I had the same issues with localtunnel not being reliable. I'll review it right away :-)

**Comment by ian-buse** [comment]
Thanks for taking a look at it. I just realized I made a small mistake with the EnhancedKeyUsageExtension critical = true. Easy fix but I wanted to point it out, I can add it to the PR.

**Review by buehler — CHANGES_REQUESTED** [review]
Hey @ian-buse 

Thanks again for the contribution!

The linter is not happy right now with something (thus, the build is failing).

A made a few comments that I'd like to discuss.

One thing about the reason: if I understand you correctly, you had problems with local tunnel being shakey (like I did). Now you implemented a special way to use self signed certificates for local development without localtunnel.

I really do like the idea! Since I had the same issues with localtunnel, I'd go even further and say we remove localtunnel altogether and make a breaking change. Then local webhooks can and should be tested with a local kubernetes instance (mini kube or whatever). What do you think about this?

I don't even know if this local tunnel feature was used heavily.

And another question: why did you replace bouncycastle? (I'm curious)

File: src/KubeOps.Operator.Web/Certificates/CertificateExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+namespace KubeOps.Operator.Web.Certificates
+{
+    public static class CertificateExtensions
+    {
+        /// <summary>
+        /// Encodes the certificate in PEM format for use in Kubernetes.
+        /// </summary>
+        /// <param name="certificate">The certificate to encode.</param>
+        /// <returns>The byte representation of the PEM-encoded certificate.</returns>
+        public static byte[] EncodeToPemBytes(this X509Certificate2 certificate) => Encoding.ASCII.GetBytes(certificate.EncodeToPem());
```
[the remark this answers is not in the document]
Comment (ian-buse) [reply src/KubeOps.Operator.Web/Certificates/CertificateExtensions.cs:15]: Yeah I think so. I believe PEM is (traditionally) encoded in ASCII, but YAML is usually UTF-8 and this is being sent as part of the YAML for the webhooks.

File: src/KubeOps.Operator.Web/Certificates/CertificateGenerator.cs [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+    /// Generates a self-signed CA certificate and server certificate using ECDsa that can be used for operator webhooks.
+    /// </summary>
+    public class CertificateGenerator : ICertificateProvider
+    {
+        private readonly string _serverName;
+        private readonly string? _serverNamespace;
+        private readonly DateTime _startDate;
+        private readonly DateTime _endDate;
+        private (X509Certificate2 Certificate, AsymmetricAlgorithm Key) _root;
+        private (X509Certificate2 Certificate, AsymmetricAlgorithm Key) _server;
```
[the remark this answers is not in the document]
Comment (ian-buse) [reply src/KubeOps.Operator.Web/Certificates/CertificateGenerator.cs:20]: Okay, I will look into doing this. Thanks.

**Comment by ian-buse** [comment]
> Hey @ian-buse
> 
> Thanks again for the contribution!
> 
> The linter is not happy right now with something (thus, the build is failing).
> 
> A made a few comments that I'd like to discuss.
> 
> One thing about the reason: if I understand you correctly, you had problems with local tunnel being shakey (like I did). Now you implemented a special way to use self signed certificates for local development without localtunnel.
> 
> I really do like the idea! Since I had the same issues with localtunnel, I'd go even further and say we remove localtunnel altogether and make a breaking change. Then local webhooks can and should be tested with a local kubernetes instance (mini kube or whatever). What do you think about this?
> 
> I don't even know if this local tunnel feature was used heavily.
> 
> And another question: why did you replace bouncycastle? (I'm curious)

Yeah, I had problems with LocalTunnel not working correctly, which made me implement the self-signed certs instead. Removing LocalTunnel would be fine with me!

BouncyCastle is a great library, but I replaced BouncyCastle with `System.Security.Cryptography` for a few reasons:

- It is easier to use with .NET classes that use certificates. For example, I can configure Kestrel like so:

    ```csharp
    using CertificateGenerator generator = new(ip);
    using X509Certificate2 cert = generator.Server.CopyServerCertWithPrivateKey();
    builder.WebHost.ConfigureKestrel(serverOptions =>
    {
        serverOptions.Listen(IPAddress.Any, 443, listenOptions =>
        {
            // If you used BouncyCastle, you would have to convert the 
            // BouncyCastle cert to X509Certificate2 before using it
            listenOptions.UseHttps(serverCert);
        });
    });
    ```

- Everything that is being done with BouncyCastle in `KubeOps.Cli` can be done with `System.Security.Cryptography`, meaning I don't think the dependency is needed. Also, some organizations would prefer the operator use the built-in .NET library over a 3rd-party library when dealing with security. I've seen other projects also moving this direction for the same reasons (web-push-libs/web-push-csharp/pull/75, planetarium/libplanet/issues/66)
- `System.Security.Cryptography` uses native DLLs and it is probably faster, though I haven't tested it myself.

I'll work on resolving the linter problem and other things you brought up. 😊

-Ian

File: src/KubeOps.Operator.Web/Certificates/CertificateGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+                key?.Dispose();
+                cert?.Dispose();
+                throw;
+            }
+        }
+
+        /// <summary>
+        /// Custom class for implementing a slim version of the .NET7/8 X509AuthorityKeyIdentifierExtension class.
+        /// </summary>
+        private sealed class CustomX509AuthorityKeyIdentifierExtension(X509Certificate2 certificate, bool critical)
```
[the remark this answers is not in the document]
Comment (buehler) [reply src/KubeOps.Operator.Web/Certificates/CertificateGenerator.cs:205]: Oh wow, good call. Actually, I never had it on the radar. Well, then we can let it slip and drop it in November.

**Comment by buehler** [comment]
Thanks for the explanation :-)

Yeah, I can absolutely understand that. I think more or less the same way, but BouncyCastle is (sometimes) great because the abstract the weird stuff like "OiD(183459122)" with something that has a meaning. That was the main reason.

I do like the move! Especially because of the native libs.

Regarding the removal of localtunnel: wdyt about this: you mark it as "deprecated" with the compiler attribute and we remove it in a later stage.

and regarding the "lazy" thing, I just tried it with the following example:
```csharp
global using CertificatePair =
    (System.Security.Cryptography.X509Certificates.X509Certificate2 Certificate,
    System.Security.Cryptography.AsymmetricAlgorithm Key);

internal class CertificateGenerator : IDisposable
{
    public Lazy<CertificatePair> Root => new(Generate);

    private CertificatePair Generate()
    {
        return default;
    }

    public void Dispose()
    {
        if (Root.IsValueCreated)
        {
            Root.Value.Certificate.Dispose();
            Root.Value.Key.Dispose();
        }
    }
}
```

What do you think about such an implementation? (just a suggestion)

**Comment by ian-buse** [comment]
> Thanks for the explanation :-)
> 
> Yeah, I can absolutely understand that. I think more or less the same way, but BouncyCastle is (sometimes) great because the abstract the weird stuff like "OiD(183459122)" with something that has a meaning. That was the main reason.
> 
> I do like the move! Especially because of the native libs.
> 
> Regarding the removal of localtunnel: wdyt about this: you mark it as "deprecated" with the compiler attribute and we remove it in a later stage.
> 
> and regarding the "lazy" thing, I just tried it with the following example:
> 
> ```cs
> global using CertificatePair =
>     (System.Security.Cryptography.X509Certificates.X509Certificate2 Certificate,
>     System.Security.Cryptography.AsymmetricAlgorithm Key);
> 
> internal class CertificateGenerator : IDisposable
> {
>     public Lazy<CertificatePair> Root => new(Generate);
> 
>     private CertificatePair Generate()
>     {
>         return default;
>     }
> 
>     public void Dispose()
>     {
>         if (Root.IsValueCreated)
>         {
>             Root.Value.Certificate.Dispose();
>             Root.Value.Key.Dispose();
>         }
>     }
> }
> ```
> 
> What do you think about such an implementation? (just a suggestion)

Good, I'm glad you like the changes! I've resolved most of the issues we talked about, including the UTF-8 encoding, linting issues, moving `ICertificateProvider` to `KubeOps.Abstractions`, and deprecating LocalTunnel. 

I like the idea of using Lazy, and I implemented a variation of your code in the latest commit. This cleaned up the interface implementation a lot. We still needed to use a backing field, otherwise the certificate will regenerate every time the property is accessed, so I followed [this Microsoft doc](https://learn.microsoft.com/en-us/dotnet/framework/performance/lazy-initialization#implementing-a-lazy-initialized-property) for how to do it.

File: src/KubeOps.Operator.Web/Certificates/ICertificateProvider.cs [hunk 4]
```diff
@@ -0,0 +1,22 @@
+﻿using System.Security.Cryptography;
+using System.Security.Cryptography.X509Certificates;
+
+namespace KubeOps.Operator.Web.Certificates
+{
+    /// <summary>
+    /// Defines properties for certificates and keys so a custom certificate/key provider may be implemented.
+    /// The provider is used by the <see cref="CertificateWebhookService"/> to provide a caBundle to the webhooks.
+    /// </summary>
+    public interface ICertificateProvider : IDisposable
```
[the remark this answers is not in the document]
Comment (ian-buse) [reply src/KubeOps.Operator.Web/Certificates/ICertificateProvider.cs]: This has been moved to KubeOps.Abstractions.

File: src/KubeOps.Operator.Web/Certificates/CertificateExtensions.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ian-buse) [reply src/KubeOps.Operator.Web/Certificates/CertificateExtensions.cs]: Done!

File: src/KubeOps.Operator.Web/Certificates/CertificateGenerator.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ian-buse) [reply src/KubeOps.Operator.Web/Certificates/CertificateGenerator.cs]: I put this in using a private field `Lazy<CertificatePair>` backing the public property `CertificatePair`.

**Review by buehler — APPROVED** [review]
Very cool :)

Just realized that the local tunnel stuff already was in "preview". As such, we can remove it later on without having a breaking change.

**Pull request merged**